api/src/main/java/org/apache/iceberg/util/CharSequenceSet.java

```diff
@@ -139,16 +139,26 @@ public boolean addAll(Collection<? extends CharSequence> charSequences) {
   @Override
   public boolean retainAll(Collection<?> objects) {
     if (objects != null) {
-      return Iterables.removeAll(wrapperSet, objects);
+      Set<CharSequenceWrapper> toRetain =
+          objects.stream()
+              .filter(CharSequence.class::isInstance)
+              .map(CharSequence.class::cast)
+              .map(CharSequenceWrapper::wrap)
+              .collect(Collectors.toSet());
+
+      return Iterables.retainAll(wrapperSet, toRetain);
     }
+
     return false;
   }
 
   @Override
+  @SuppressWarnings("CollectionUndefinedEquality")
   public boolean removeAll(Collection<?> objects) {
     if (objects != null) {
-      return Iterables.removeAll(wrapperSet, objects);
+      return objects.stream().filter(this::remove).count() != 0;
```

Comment:
why not do the same as in L142ff?

Reply:
`removeAll` iterates the `objects` and checks whether these elements exist in the `removeFrom` set, in this case we don't need to wrap all elements at once, use `this#remove` can take advantage of the wrapper in the `ThreadLocal` `wrappers`.